Ability to specify custom filters in filterQuery

vonagam opened the pull request:
feathers-mongoose have `$populate` which is non-common api filter.

Other database wrappers might have extra filters too (for example feathers-objection have `$eager` which is similar in effect to `$populate`).

This pull allows specifying such db specific filters as third argument for `filterQuery`.

vonagam committed:
New version of filterQuery

daffl commented:
Thank you for the pull request! Since this repository is mostly an internal utility library, how would this be used?

vonagam commented:
If it is considered internal to feathers org repositories then you can close it, because you can manually add specific db filters to the list (as with $populate). And other stuff in the repository looks like really internal.

But `filterQuery` looks more like common utility for any database service which want to support feathers querying api.

In my case as you remember in my `$and` pull, i wanted to merge two queries. One internal and one coming from user. I can do it like this `{$and: [userQuery, restrictionQuery]}`, but user query can contain filters (select, order, and other stuff), so what is right to do is something like this: `{$and: [userQuery, restrictionQuery], ...userFilters}`. To do separation i use `filterQuery`. But i use `feathers-objection` which supports additional filters, so i have to wrap `filterQuery` to handle those too.

Is there alternative function which considered not internal, which handles separation of query and filters?

daffl commented:
Oh, I didn't mean that it should be closed, I was just pointing out that all packages that rely on this have to be updated accordingly since this module is usually not something any Feathers user would use directly.

What I would really like to do for the next versions of the db adapters is similar to your change actually. For security reasons it makes a lot of sense to recursively filter out all `$` prefixed key names unless they are part of the official syntax or explicitly whitelisted. Maybe we can tackle this here together.

vonagam commented:
By converting operators do you mean replacing `"$or"` with `Op.or`?
Seems like too specific thing that should take place in post processing of cleaned query.
I mean how many wrappers do need this sorts of functionality (is there any other)?
And it's not like not providing it in `filterQuery` will make it any harder to do.

daffl commented:
True. I think this makes sense then. Do you want to update your PR? I'll can make sure it gets included in the next versions of all database adapters then.

vonagam commented:
Updated. Removed default `$populate` filter. Should be added through `filters` option in mongoose adapter. (Can add back to defaults, if for some reason it is more preferable.)

By the way, why `$and` operator is not included in default ones?

daffl commented on test/filter-query.test.js:
```diff
@@ -3,7 +3,7 @@ const filter = require('../lib/filter-query');
 const expect = chai.expect;
 
-describe('.filterQuery', function () {
+describe.only('.filterQuery', function () {
```
We do want to run all the test 😉

daffl reviewed (commented):
Awesome, thank you! Just one last small thing. The default query parameter composition is through `$and` which is why it isn't part of the standard query syntax - and most database adapters. With these changes we can always still add it to the adapters that do support it.

daffl commented:
Yeah, it can get kind of complicated for non-MongoDB adapters but we can track it in a separate issue and revisit later. Going to merge this one and it will go out in the next (major) version. Thank you again for doing this!

sean-nicholas commented:
What about allowing non-feathers operators only for certain fields? Here is a test from my own implementation of such a behavior:

```js
test('should ignore mongo operators if whitelisted deep', () => {
    const result = hasMongoOperators(wrapInContext({
      test: '123',
      second: [
        {
          $regex: '.*'
        }
      ]
    }), { whitelist: ['second.0.$regex'] })
    expect(result).toEqual(false)
  })
```

This would allow a more granular whitelist.

Another thing: Is filtering the right way or should an exception be thrown? I would prefer the former one. I wrote a blog post about it: https://medium.com/@sean.nicholas/opinion-dont-mutate-the-input-or-output-of-your-services-most-of-the-time-e764047c8c96 It needs some revision because some people did not get what I want to say: Silently removing parts from users input (here the query) can be really frustrating for the frontend developer. "Why does my query return the wrong data?" vs. "Oh there is an error: Operator XY is not allowed. So that needs to be fixed in the backend."
